src/data-model-providers/codegen/EmberAttributeStorageImpl.cpp

```diff
@@ -0,0 +1,164 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "lib/core/CHIPError.h"
+#include <data-model-providers/codegen/EmberAttributeStorageImpl.h>
+
+#include <app-common/zap-generated/attribute-type.h>
+#include <app/util/persistence/AttributePersistenceProvider.h>
+#include <lib/support/ScopedBuffer.h>
+
+namespace chip {
+namespace app {
+namespace Storage {
+namespace {
+
+// EmberAfAttributeType possible values are part of the unnamed enum in `attribute-type.h`,
+// of the form ZCL_*_ATTRIBUTE_TYPE
+EmberAfAttributeType attributeType(AttributeStorage::Buffer & buffer)
+{
+    switch (buffer.type())
+    {
+    case AttributeStorage::Buffer::Type::kPrimitive:
+        switch (buffer.size())
+        {
+        case 1:
+            return ZCL_INT8U_ATTRIBUTE_TYPE;
+        case 2:
+            return ZCL_INT16U_ATTRIBUTE_TYPE;
+        case 3:
+            return ZCL_INT24U_ATTRIBUTE_TYPE;
+        case 4:
+            return ZCL_INT32U_ATTRIBUTE_TYPE;
+        case 5:
+            return ZCL_INT40U_ATTRIBUTE_TYPE;
+        case 6:
+            return ZCL_INT48U_ATTRIBUTE_TYPE;
+        case 7:
+            return ZCL_INT56U_ATTRIBUTE_TYPE;
+        case 8:
+            return ZCL_INT64U_ATTRIBUTE_TYPE;
+        default:
+            // This one is a bit nonsense..
+            return ZCL_NO_DATA_ATTRIBUTE_TYPE;
+        }
+    case AttributeStorage::Buffer::Type::kMutableByteSpan:
+        // Generally ember persistence did not persist raw values.
+        return ZCL_STRUCT_ATTRIBUTE_TYPE;
+    case AttributeStorage::Buffer::Type::kStringOneByteLength:
+    case AttributeStorage::Buffer::Type::kBytesOneByteLength:
+        return ZCL_OCTET_STRING_ATTRIBUTE_TYPE;
+    case AttributeStorage::Buffer::Type::kStringTwoByteLength:
+    case AttributeStorage::Buffer::Type::kBytesTwoByteLength:
+        return ZCL_LONG_OCTET_STRING_ATTRIBUTE_TYPE;
+    }
+
+    // This should NEVER be reached, however I did have compilers ocmplain about it.
+    return ZCL_NO_DATA_ATTRIBUTE_TYPE;
+}
+
+} // namespace
+
+CHIP_ERROR EmberAttributeStorageImpl::Write(const ConcreteAttributePath & path, const Value & value)
+{
+    AttributePersistenceProvider * provider = GetAttributePersistenceProvider();
+
+    // note that data is already in the format that EMBER functions expect:
+    //  - numeric is "as is"
+    //  - pascal strings are already length-prefixed
+    //  - raw values are just bytes
+    return provider->WriteValue(path, value.data());
+}
+
+CHIP_ERROR EmberAttributeStorageImpl::Read(const ConcreteAttributePath & path, Buffer buffer)
+{
+    VerifyOrReturnError(buffer.size() > 0, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    AttributePersistenceProvider * provider = GetAttributePersistenceProvider();
+    const EmberAfAttributeMetadata fakeMetadata{
```

Comment:
Please comment here that in practice AttributePersistenceProvider only cares about size and attributeType here?

But nothing actually enforces or documents that right now, so maybe we should update the API docs on that and add upgrade notes?